File: backend/src/main/java/com/parapharma/service/ProductService.java
package com.parapharma.service;

import com.parapharma.dto.ProductCreateDTO;
import com.parapharma.dto.ProductDTO;
import com.parapharma.dto.ProductFilterDTO;
import com.parapharma.entity.Category;
import com.parapharma.entity.Product;
import com.parapharma.entity.ProductImage;
import com.parapharma.exception.ResourceNotFoundException;
import com.parapharma.mapper.ProductMapper;
import com.parapharma.repository.CategoryRepository;
import com.parapharma.repository.ProductRepository;
import com.parapharma.repository.ProductSpecification;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.text.Normalizer;
import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional
@Slf4j
public class ProductService {

    private final ProductRepository productRepository;
    private final CategoryRepository categoryRepository;
    private final ProductMapper productMapper;
    private final FileStorageService fileStorageService;

    @Transactional(readOnly = true)
    public Page<ProductDTO> getProducts(ProductFilterDTO filter) {
        Sort sort = Sort.by(
                filter.getSortDir().equalsIgnoreCase("asc") ? Sort.Direction.ASC : Sort.Direction.DESC,
                filter.getSortBy());
        Pageable pageable = PageRequest.of(filter.getPage(), filter.getSize(), sort);

        Page<Product> products = productRepository.findAll(
                ProductSpecification.withFilters(
                        filter.getKeyword(),
                        filter.getCategoryId(),
                        filter.getManufacturer(),
                        filter.getMinPrice(),
                        filter.getMaxPrice(),
                        filter.getMinRating(),
                        filter.getInStock()),
                pageable);

        return products.map(productMapper::toDTO);
    }

    @Transactional(readOnly = true)
    public ProductDTO getProductBySlug(String slug) {
        Product product = productRepository.findBySlug(slug)
                .orElseThrow(() -> new ResourceNotFoundException("Produit", "slug", slug));
        return productMapper.toDTO(product);
    }

    @Transactional(readOnly = true)
    public ProductDTO getProductById(Long id) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Produit", "id", id));
        return productMapper.toDTO(product);
    }

    @Transactional(readOnly = true)
    public Page<ProductDTO> searchProducts(String query, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        return productRepository.search(query, pageable).map(productMapper::toDTO);
    }

    @Transactional(readOnly = true)
    public List<ProductDTO> getRelatedProducts(String slug, int limit) {
        Product product = productRepository.findBySlug(slug)
                .orElseThrow(() -> new ResourceNotFoundException("Produit", "slug", slug));
        Pageable pageable = PageRequest.of(0, limit);
        List<Product> related = productRepository.findRelatedProducts(
                product.getCategory().getId(), product.getId(), pageable);
        return productMapper.toDTOList(related);
    }

    @Transactional(readOnly = true)
    public List<String> getAllManufacturers() {
        return productRepository.findAllManufacturers();
    }

    @Transactional(readOnly = true)
    public Page<ProductDTO> getLatestProducts(int page, int size) {
        return productRepository.findLatestProducts(PageRequest.of(page, size))
                .map(productMapper::toDTO);
    }

    @Transactional(readOnly = true)
    public Page<ProductDTO> getTopRatedProducts(int page, int size) {
        return productRepository.findTopRatedProducts(PageRequest.of(page, size))
                .map(productMapper::toDTO);
    }

    public ProductDTO createProduct(ProductCreateDTO dto) {
        Category category = categoryRepository.findById(dto.getCategoryId())
                .orElseThrow(() -> new ResourceNotFoundException("Catégorie", "id", dto.getCategoryId()));

        String slug = generateSlug(dto.getTitle());

        Product product = Product.builder()
                .title(dto.getTitle())
                .slug(slug)
                .price(dto.getPrice())
                .rating(dto.getRating() != null ? dto.getRating() : 0.0)
                .description(dto.getDescription())
                .usageTips(dto.getUsageTips())
                .manufacturer(dto.getManufacturer())
                .category(category)
                .stockQuantity(dto.getStockQuantity() != null ? dto.getStockQuantity() : 0)
                .active(dto.getActive() != null ? dto.getActive() : true)
                .build();

        Product saved = productRepository.save(product);
        return productMapper.toDTO(saved);
    }

    public ProductDTO updateProduct(Long id, ProductCreateDTO dto) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Produit", "id", id));

        Category category = categoryRepository.findById(dto.getCategoryId())
                .orElseThrow(() -> new ResourceNotFoundException("Catégorie", "id", dto.getCategoryId()));

        product.setTitle(dto.getTitle());
        product.setPrice(dto.getPrice());
        product.setDescription(dto.getDescription());
        product.setUsageTips(dto.getUsageTips());
        product.setManufacturer(dto.getManufacturer());
        product.setCategory(category);
        product.setStockQuantity(dto.getStockQuantity() != null ? dto.getStockQuantity() : product.getStockQuantity());
        product.setRating(dto.getRating() != null ? dto.getRating() : product.getRating());
        product.setActive(dto.getActive() != null ? dto.getActive() : product.getActive());

        if (!product.getTitle().equals(dto.getTitle())) {
            product.setSlug(generateSlug(dto.getTitle()));
        }

        Product saved = productRepository.save(product);
        return productMapper.toDTO(saved);
    }

    public void deleteProduct(Long id) {
        if (!productRepository.existsById(id)) {
            throw new ResourceNotFoundException("Produit", "id", id);
        }
        productRepository.deleteById(id);
    }

    public ProductDTO addImageToProduct(Long productId, String imageUrl, boolean isPrimary) {
        log.info("Liaison de l'image {} au produit ID: {} (Primaire: {})", imageUrl, productId, isPrimary);
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new ResourceNotFoundException("Produit", "id", productId));

        if (isPrimary) {
            log.debug("Réinitialisation des images primaires pour le produit : {}", productId);
            product.getImages().forEach(img -> img.setIsPrimary(false));
        }

        ProductImage image = ProductImage.builder()
                .imageUrl(imageUrl)
                .displayOrder(product.getImages().size())
                .isPrimary(isPrimary)
                .build();

        product.addImage(image);
        Product saved = productRepository.save(product);
        log.info("Image ajoutée avec succès. Nombre total d'images : {}", saved.getImages().size());
        return productMapper.toDTO(saved);
    }

    public ProductDTO removeImageFromProduct(Long productId, Long imageId) {
        log.info("Suppression de l'image ID: {} du produit ID: {}", imageId, productId);
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new ResourceNotFoundException("Produit", "id", productId));

        ProductImage imageToRemove = product.getImages().stream()
                .filter(img -> img.getId().equals(imageId))
                .findFirst()
                .orElseThrow(() -> new ResourceNotFoundException("Image", "id", imageId));

        // Delete the physical file from disk
        fileStorageService.deleteFile(imageToRemove.getImageUrl());

        product.removeImage(imageToRemove);
        Product saved = productRepository.save(product);
        log.info("Image supprimée de la base de données. ID: {}", imageId);
        return productMapper.toDTO(saved);
    }

    private String generateSlug(String title) {
        String normalized = Normalizer.normalize(title, Normalizer.Form.NFD)
                .replaceAll("[\\p{InCombiningDiacriticalMarks}]", "");
        String slug = normalized.toLowerCase()
                .replaceAll("[^a-z0-9\\s-]", "")
                .replaceAll("\\s+", "-")
                .replaceAll("-+", "-")
                .replaceAll("^-|-$", "");

        // Ensure uniqueness
        String baseSlug = slug;
        int counter = 1;
        while (productRepository.existsBySlug(slug)) {
            slug = baseSlug + "-" + counter;
            counter++;
        }
        return slug;
    }
}

File: frontend/src/app/core/services/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface Product {
  id: number;
  title: string;
  slug: string;
  price: number;
  rating: number;
  description: string;
  usageTips: string;
  manufacturer: string;
  categoryId: number;
  categoryName: string;
  categorySlug: string;
  images: ProductImage[];
  stockQuantity: number;
  active: boolean;
  createdAt: string;
}

export interface ProductImage {
  id: number;
  imageUrl: string;
  displayOrder: number;
  isPrimary: boolean;
}

export interface ProductPage {
  content: Product[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
  first: boolean;
  last: boolean;
}

export interface Category {
  id: number;
  name: string;
  slug: string;
  description: string;
  imageUrl: string;
  productCount?: number;
}

export interface ProductFilter {
  keyword?: string;
  categoryId?: number;
  manufacturer?: string;
  minPrice?: number;
  maxPrice?: number;
  minRating?: number;
  inStock?: boolean;
  sortBy?: string;
  sortDir?: string;
  page?: number;
  size?: number;
}

@Injectable({ providedIn: 'root' })
export class ProductService {
  private apiUrl = `${environment.apiUrl}/products`;
  private catUrl = `${environment.apiUrl}/categories`;

  constructor(private http: HttpClient) {}

  getProducts(filter: ProductFilter = {}): Observable<ProductPage> {
    let params = new HttpParams();
    if (filter.keyword) params = params.set('keyword', filter.keyword);
    if (filter.categoryId) params = params.set('categoryId', filter.categoryId.toString());
    if (filter.manufacturer) params = params.set('manufacturer', filter.manufacturer);
    if (filter.minPrice !== undefined) params = params.set('minPrice', filter.minPrice.toString());
    if (filter.maxPrice !== undefined) params = params.set('maxPrice', filter.maxPrice.toString());
    if (filter.minRating !== undefined) params = params.set('minRating', filter.minRating.toString());
    if (filter.inStock !== undefined) params = params.set('inStock', filter.inStock.toString());
    if (filter.sortBy) params = params.set('sortBy', filter.sortBy);
    if (filter.sortDir) params = params.set('sortDir', filter.sortDir);
    params = params.set('page', (filter.page || 0).toString());
    params = params.set('size', (filter.size || 12).toString());

    return this.http.get<ProductPage>(this.apiUrl, { params });
  }

  getProductBySlug(slug: string): Observable<Product> {
    return this.http.get<Product>(`${this.apiUrl}/${slug}`);
  }

  searchProducts(query: string, page = 0, size = 12): Observable<ProductPage> {
    const params = new HttpParams()
      .set('q', query)
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<ProductPage>(`${this.apiUrl}/search`, { params });
  }

  getRelatedProducts(slug: string, limit = 4): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/${slug}/related?limit=${limit}`);
  }

  getLatestProducts(page = 0, size = 8): Observable<ProductPage> {
    const params = new HttpParams().set('page', page.toString()).set('size', size.toString());
    return this.http.get<ProductPage>(`${this.apiUrl}/latest`, { params });
  }

  getTopRatedProducts(page = 0, size = 8): Observable<ProductPage> {
    const params = new HttpParams().set('page', page.toString()).set('size', size.toString());
    return this.http.get<ProductPage>(`${this.apiUrl}/top-rated`, { params });
  }

  getManufacturers(): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/manufacturers`);
  }

  getCategories(): Observable<Category[]> {
    return this.http.get<Category[]>(this.catUrl);
  }

  // Admin methods
  createProduct(product: any): Observable<Product> {
    return this.http.post<Product>(this.apiUrl, product);
  }

  updateProduct(id: number, product: any): Observable<Product> {
    return this.http.put<Product>(`${this.apiUrl}/${id}`, product);
  }

  deleteProduct(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  uploadImage(productId: number, file: File, primary = false): Observable<Product> {
    const formData = new FormData();
    formData.append('file', file);
    return this.http.post<Product>(`${this.apiUrl}/${productId}/images?primary=${primary}`, formData);
  }

  deleteImage(productId: number, imageId: number): Observable<Product> {
    return this.http.delete<Product>(`${this.apiUrl}/${productId}/images/${imageId}`);
  }
}
